gencred/merge_kubeconfig_secret.py

```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env python3
+
+# Copyright 2020 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Merges a kubeconfig file into a kubeconfig file in a k8s secret giving precedence to the secret."""
+
+# Requirements: kubectl (pointed at the correct cluster) and base64
+# Example usage:
+# ./merge_kubeconfig_secret.py \
+#   --secret-name=mykube \
+#   --secret-namespace=myspace \
+#   --secret-key=data-key \
+#   kubeconfig.yaml
+
+import argparse
+import os
+import subprocess
+import sys
+import tempfile
+
+def call(cmd, **kwargs):
+	print('>>> %s' % cmd)
+	return subprocess.run(
+		cmd,
+		check=True,
+		shell=True,
+		stderr=sys.stderr,
+		stdout=subprocess.PIPE,
+		timeout=10,#seconds
+		universal_newlines=True,
+		**kwargs,
+	)
+
+def main(args):
+	print(args)
+	with tempfile.TemporaryDirectory() as tmpdir:
+		orig = '%s/original' % (tmpdir)
+		merged = '%s/merged' % (tmpdir)
+		# Copy the current secret contents into a temp file.
+		cmd = 'kubectl get secret --namespace "%s" "%s" -o go-template="{{index .data \\"%s\\"}}" | base64 -d > %s' % (args.secret_namespace, args.secret_name, args.secret_key, orig)
+		call(cmd)
+
+		# Merge the existing and new kubeconfigs into another temp file.
+		env = os.environ.copy()
+		env['KUBECONFIG'] = '%s:%s' % (orig, args.kubeconfig_to_merge)
+		call(
+			'kubectl config view --raw > %s' % (merged),
+			env=env,
+		)
+
+		# Update the secret with the merged config.
+		call('kubectl create secret generic --namespace "%s" "%s" --from-file="%s=%s" --dry-run -oyaml | kubectl apply -f -' % (args.secret_namespace, args.secret_name, args.secret_key, merged))
+
+		print('Successfully updated secret %s/%s.' % (args.secret_namespace, args.secret_name))
+
+if __name__ == '__main__':
+	parser = argparse.ArgumentParser(description='Merges the provided kubeconfig file into a kubeconfig file living in a kubernetes secret in order to add new cluster contexts to the secret. Requires kubectl and base64.')
+	parser.add_argument(
+	    '--secret-name',
+	    help='The name of the k8s secret containing the kubeconfig file to add to.',
+	    default='kubeconfig',
+	)
+	parser.add_argument(
+	    '--secret-namespace',
```

Comment:
Do we really need the `--secret` prefix here and key? Seems redundant given the scope.